cores/arduino/wiring_digital.c

```diff
@@ -51,18 +37,16 @@ void pinMode(pin_size_t ulPin, PinMode ulMode)
         case INPUT_PULLDOWN:
             pin_function(p, GD_PIN_FUNCTION3(PIN_MODE_IPD, 0, 0));
             break;
- // now done in analogRead().
- //       case INPUT_ANALOG:
- //           pin_function(p, GD_PIN_FUNCTION3(PIN_MODE_AIN, 0, 0));
- //           break;
         case OUTPUT:
             pin_function(p, GD_PIN_FUNCTION3(PIN_MODE_OUT_PP, 0, 0));
             break;
- //       case OUTPUT_OPEN_DRAIN:
- //           pin_function(p, GD_PIN_FUNCTION3(PIN_MODE_OUT_OD, 0, 0));
- //           break;
+        case INPUT_ANALOG: // From PinModeExtension
+            pin_function(p, GD_PIN_FUNCTION3(PIN_MODE_AIN, 0, 0));
+            break;
+        case OUTPUT_OPEN_DRAIN: // From PinModeExtension
+            pin_function(p, GD_PIN_FUNCTION3(PIN_MODE_OUT_OD, 0, 0));
```

Comment:
While the `GD_PIN_FUNCTION3(PIN_MODE_OUT_OD, 0, 0)` works for the GD32F30x series, the macro definition is [actually](https://github.com/CommunityGD32Cores/ArduinoCore-GD32/blob/e9bdbef9ff1bc89671afd3eb5a82eb34c7311f30/cores/arduino/gd32/PinNames.h#L315)

```
#define GD_PIN_FUNCTION3(MODE, ODPP, AFN) ((int)(MODE & PIN_MODE_MASK) | ((ODPP & PIN_OUTPUT_MODE_MASK) << PIN_OUTPUT_MODE_SHIFT) | ((AFN & PIN_AF_MASK) << PIN_AF_SHIFT))
```

```c
enum {
    PIN_OTYPE_PP           = 0,
    PIN_OTYPE_OD           = 1,
};
``` 
So the second macro parameter shouldn't be 0 (push-pull output mode) but `PIN_OTYPE_OD`. That output-mode part of the macro is used in the GD32F4xx path of [`pin_function`](https://github.com/CommunityGD32Cores/ArduinoCore-GD32/blob/34a0f1c90240af32f3271b71579570582b042831/cores/arduino/gd32/pinmap.c#L68-L86) since SPL has a different API format there.

This core doesn't currently support GD32F4xx, but we'd future-proof it a bit if we add the correct code now, and it won't affect GD32F30x since the variable goes unused in that code path.

Reply:
Huh. This was code directly from the original code from GigaDevice.  But really shouldn't we replace the 0 and 1 with the enum keys PIN_OTYPE_OD and PIN_OTYPE_PP?

Reply:
They likely didn't test it against GD32F4xx because the core doesn't compile for it as a whole. I'd propose using 0 where the output mode does not matter (because the pin mode is input or analog), and the explicit PushPull or OpenDrain macro where the pin mode would be respected (in GD32F4xx), just like you did in the recent commit.